src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,304 @@
+    debug!("unused_generic_params: (after visitor) used_parameters={:064b}", used_parameters);
+
+    mark_used_by_predicates(tcx, def_id, &mut used_parameters);
+    debug!("unused_generic_params: (after predicates) used_parameters={:064b}", used_parameters);
+
+    // Invert the u64 so that used is 0 and unused is 1. This makes checking if all parameters are
+    // used easy - just compare with zero.
+    debug!("unused_generic_params: (end) used_parameters={:064b}", used_parameters);
+    let unused_parameters: u64 = !used_parameters;
+    debug!("unused_generic_params: (flipped) unused_parameters={:064b}", unused_parameters);
+
+    // Emit errors for debugging and testing if enabled.
+    let is_full = unused_parameters == 0;
+    if tcx.sess.opts.debugging_opts.polymorphize_errors && !is_full {
+        emit_unused_generic_params_error(tcx, def_id, generics, unused_parameters);
+    }
+
+    unused_parameters
+}
+
+/// Checks if the `param_index`th bit is set (or out-of-range).
+fn is_bit_set(parameters: u64, param_index: u32) -> bool {
+    param_index >= 64 || ((parameters.checked_shr(param_index).unwrap_or(1)) & 1) == 1
+}
+
+/// Flips the bit corresponding to the parameter index.
+fn set_bit(used_parameters: &mut u64, param_index: u32) {
+    debug!("set_bit: used_parameters={:064b} param_index={:?}", used_parameters, param_index);
+    *used_parameters |= 1u64.checked_shl(param_index).unwrap_or(0);
+    debug!("set_bit: used_parameters={:064b}", used_parameters);
+}
+
+/// Some parameters are considered used-by-default, such as non-generic parameters and the dummy
+/// generic parameters from closures, this function marks them as used. `leaf_is_closure` should
+/// be `true` if the item that `unused_generic_params` was invoked on is a closure.
+fn mark_used_by_default_parameters<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    used_parameters: &mut u64,
+) {
+    if !tcx.is_trait(def_id) && (tcx.is_closure(def_id) || tcx.type_of(def_id).is_generator()) {
+        for param in &generics.params {
+            debug!("mark_used_by_default_parameters: (closure/gen) param={:?}", param);
+            set_bit(used_parameters, param.index);
+        }
+    } else {
+        for param in &generics.params {
+            debug!("mark_used_by_default_parameters: (other) param={:?}", param);
+            if let ty::GenericParamDefKind::Lifetime = param.kind {
+                set_bit(used_parameters, param.index);
+            }
+        }
+    }
+
+    if let Some(parent) = generics.parent {
+        mark_used_by_default_parameters(tcx, parent, tcx.generics_of(parent), used_parameters);
+    }
+}
+
+/// Search the predicates on used generic parameters for any unused generic parameters, and mark
+/// those as used.
+fn mark_used_by_predicates<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, used_parameters: &mut u64) {
+    // Cannot get `explicit_predicates_of` for closures.
+    if tcx.is_closure(def_id) {
```

Comment:
Presumably you can query that on `closure_base_def_id(def_id)` instead?

Reply:
Yes, fixed!